### ISSUE TITLE
insertOoxml for Word does not work on WOPI host (CSPP)

### ISSUE THREAD

**Issue by nickamckenna** [OfficeDev/office-js#2893]
# Provide required information needed to triage your issue
I've attached a very simple demo Add-In which works fine in desktop and O365, but fails in the WOPI Word editor provided by CSPP (the Microsoft Cloud Storage Partner Program).

These lines of code:

        const range = context.document.getSelection();
        range.clear();
        range.insertOoxml(dummyXml, Word.InsertLocation.replace, false);

always result in this error in the console (and the Ooxml is not inserted):

App.js:42 RichApi.Error: Cannot read properties of null (reading 'message')
    at new n (word-web-16.00.js:25:324945)
    at o.processRequestExecutorResponseMessage (word-web-16.00.js:25:389111)
    at word-web-16.00.js:25:387174

## Your Environment

* Platform [PC desktop, Mac, iOS, Office on the web]: WOPI editor provided by CSPP
* Host [Excel, Word, PowerPoint, etc.]: Word
* Office version number: 16.0.15810.41012
* Operating System: Any
* Browser (if using Office on the web): Chrome
 
## Expected behavior

The Ooxml should be inserted as happens with desktop and O365.

## Current behavior

The text is not inserted and an error is logged in the console.

App.js:42 RichApi.Error: Cannot read properties of null (reading 'message')
    at new n (word-web-16.00.js:25:324945)
    at o.processRequestExecutorResponseMessage (word-web-16.00.js:25:389111)
    at word-web-16.00.js:25:387174

## Steps to reproduce

1. Use the sample code attached with the WOPI editor provided by CSPP
2. Click the "Insert OOXML" button
3. Note that text is not inserted
4. Note that the error appears in the console

## Link to live example(s)

1. Download sample code attached
2. ______
3. ______

# Provide additional details

1. ______
2. ______
3. ______

## Context

We are inserting text from a library in our Add-In. We are trying to do this with Content Controls, but the problem seems to occur even with simple ranges or any use of insertOoxml.
[Ooxml Bug Demo.zip](https://github.com/OfficeDev/office-js/files/9754057/Ooxml.Bug.Demo.zip)

## Useful logs

- [ ] Console errors
- [ ] Screenshots
- [ ] Test file (if only happens on a particular file)

Test Add-In attached on "Context" above.

Error in console:

App.js:42 RichApi.Error: Cannot read properties of null (reading 'message')
    at new n (word-web-16.00.js:25:324945)
    at o.processRequestExecutorResponseMessage (word-web-16.00.js:25:389111)
    at word-web-16.00.js:25:387174

Thank you for taking the time to report an issue. Our triage team will respond to you in less than 72 hours. Normally, response time is <10 hours Monday through Friday. We do not triage on weekends.

**Comment by RuizhiSunMS** [comment]
Thanks for your submitting. 
This issue is somewhat complicated, and we decide to create a bug #6936950 to track.
Sorry for unable to provide a precise timeline now but will contact you if any update appears.

**Comment by nickamckenna** [comment]
Thanks - how do I track progress on this bug? Is there a link to it somewhere?

**Comment by nickamckenna** [comment]
Is there an update on this? How can I view the status of bug #6936950?

**Comment by kommandor** [comment]
Are there any updates? We are also waiting for that bug resolving.

**Comment by nickamckenna** [comment]
I have heard nothing! There is a bug number (#6936950) mentioned above, but no indication on how to view / track that bug!

**Comment by makc-de** [comment]
@RuizhiSunMS could you provide the status for the bug #6936950? Are there any schedule / plans regarding it?

**Comment by nickamckenna** [comment]
For anyone reading this in future, context.document.insertFileFromBase64(...) does not work either.

**Comment by aeturneus** [comment]
@RuizhiSunMS or anyone else, is there any status update for bug #6936950?

This is a significant issue. Has anyone figured out a workaround?

**Comment by xiaoyuMS** [comment]
@aeturneus We are fixing this issue, I will come back once it is ready. Similar issue: #3040

**Comment by xiaoyuMS** [comment]
Hi @aeturneus  @nickamckenna  @kommandor  @makc-de , the fix has been deployed to production, please have a try. The full deployment may take another 1-2 days.

**Comment by vnau** [comment]
Just tried to use InsertOoxml in WOPI in the Test/Dogfood environment, unfortunately it still doesn't work and throws errors:

`WordEditorDS.js:7544 ULS tag_bgbk4: AppId:PRIVATE, Can't find the inserted content when setting ooxml, retryTime:0
core.mjs:8400 ERROR Error: Uncaught (in promise): GeneralException: Can't find the inserted content. Please refresh the page for the result.
RichApi.Error: Can't find the inserted content. Please refresh the page for the result.
`

**Comment by SavioSantosDev** [comment]
@xiaoyuMS It still doesn't work on WOPI Host.

For us, the insertOoxml function previously inserted content, but we needed to refresh the page to display the content inserted into the document. This week it stopped working completely.
```
   Word.run(async context => {
        const cursor = context.document.getSelection();
        context.load(cursor);
        cursor.insertOoxml(ooxml, Word.InsertLocation.after);

        await context.sync();
      })
```

Basically, the function is called, Word loads and after a while an error appears in the console:
![word-loading](https://github.com/OfficeDev/office-js/assets/48107865/2dbb1a6c-66fa-42fd-a46c-ab14efdfcb36)

![console-error](https://github.com/OfficeDev/office-js/assets/48107865/22d1179e-62f6-486c-87e7-3f349c7b0461)

**Comment by aeturneus** [comment]
@xiaoyuMS we are having a similar issue in WOPI, however we now see a different error:

`{
    "code": "NotAllowed",
    "message": "The action isn't supported by Word in a browser. Check the OfficeExtension.Error.debugInfo for more information.",
    "errorLocation": "Body.insertOoxml",
    "statement": "var insertOoxml = body.insertOoxml(...);",
    "surroundingStatements": [
        "var root = context.root;",
        "var body = root.body;",
        "// Instantiate {body}",
        "// >>>>>",
        "var insertOoxml = body.insertOoxml(...);",
        "// <<<<<",
        "// Instantiate {insertOoxml}"
    ],
    "fullStatements": [
        "Please enable config.extendedErrorLogging to see full statements."
    ]
}`

this functionality is crucial to our use case so we really hope it can be used in WOPI soon.

**Comment by xiaoyu03** [comment]
Hi @SavioSantosDev @aeturneus, thanks for trying! We have noticed different failures on WOPI instances and now have an internal work item to track it. Once we have made more progress, we will return to this thread.

**Comment by aeturneus** [comment]
@xiaoyu03 anything to report? We are looking forward to launching a new feature once this issue is resolved, thanks!

**Comment by xiaoyu03** [comment]
Hi @aeturneus @SavioSantosDev @vnau we have deployed a fix for the WOPI host issue, please have a try~

**Comment by vnau** [comment]
Hi @xiaoyu03 from our side we can confirm that insertOoxml now inserts content successfully.
Thank you very much for your work, this is really a breakthrough for us.

**Comment by aeturneus** [comment]
hi @xiaoyu03 we still see the same error message:

`{ "code": "NotAllowed", "message": "The action isn't supported by Word in a browser. Check the OfficeExtension.Error.debugInfo for more information.", "errorLocation": "Body.insertOoxml", "statement": "var insertOoxml = body.insertOoxml(...);", "surroundingStatements": [ "var root = context.root;", "var body = root.body;", "// Instantiate {body}", "// >>>>>", "var insertOoxml = body.insertOoxml(...);", "// <<<<<", "// Instantiate {insertOoxml}" ], "fullStatements": [ "Please enable config.extendedErrorLogging to see full statements." ] }`

**Comment by xiruatms** [comment]
@aeturneus , will you please share more details of your environment for us to continue investigation?

**Comment by aeturneus** [comment]
@xiruatms sure. We are using a production WOPI discovery url, running our add-in in Word Online, with the following configuration: 
![image](https://github.com/OfficeDev/office-js/assets/877362/d0ee6941-c62e-48b5-83c9-9fb3f97c492b)

The operation that fails only in WOPI (but works in desktop Word) is as follows:

![image](https://github.com/OfficeDev/office-js/assets/877362/8b1cd124-a9e0-4074-879c-7daad09c7705)

Please let me know if you need any further information.

**Comment by xiruatms** [comment]
Thanks. 
Will you please provide the Word Online build number and/or screenshot?
That'll be great if you can provide us some testing accounts to access your environment.

**Comment by aeturneus** [comment]
@xiruatms our Word Online build number is 16.0.17508.41004

In our application, we use insertOoxml to toggle (insert/remove) sections of a document. Here is a screenshot 
<img width="1629" alt="image" src="https://github.com/OfficeDev/office-js/assets/877362/2966b1a1-55f8-473b-93b4-aa5e1b672733">

Let me know how we can send you credentials for testing to access our production environment.

**Comment by aeturneus** [comment]
@xiruatms seeing same problem on 16.0.17514.41008 build of Word Online

If there's anything we can do to move forward with this let me know

![image](https://github.com/OfficeDev/office-js/assets/877362/426e27ca-bb53-41d1-9aac-c180618edd07)

**Comment by SavioSantosDev** [comment]
@Somyo5941 Here for us, the insertOoxml function is working again, but with the same problem we had before, of having to refresh the page to view the inserted content. In fact, the first time it is called it works normally, but from the second time onwards it causes problems:

![test_addin-2024-03-18_14 08 03](https://github.com/OfficeDev/office-js/assets/48107865/1a78e167-78c2-40d4-a4fe-4073086d0684)

Is this a known issue? Is there anything that can be done to avoid having to refresh the page?

**Comment by xiruatms** [comment]
@aeturneus , will you please provide us more details about your environment using following survey? Please provide URL, testing accounts, testing documents and so on so that we can look into more details. Thanks

https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbR9NvrlmrippKsMvh4kTM325URVZMWjYyQTVOTVNLTUo5VzFRTTcyNU1RSy4u

**Comment by aeturneus** [comment]
@xiruatms thanks, I have submitted this form with credentials and detailed steps to reproduce.

**Comment by xiaoyu03** [comment]
Hi @SavioSantosDev , sorry for the late response. Could you clarify more about the details? Does it work on other insertion cases, such as insertion in a new paragraph? Or it's just that it will fail the second time.

**Comment by xiaoyuMS** [comment]
Hi @SavioSantosDev , sorry for the late response. Could you clarify more about the details? Does it work on other insertion cases, such as insertion in a new paragraph? Or it's just that it will fail the second time.

**Comment by SavioSantosDev** [comment]
Hi @xiaoyuMS `insertParagraph` and `insertText` work normally, but unfortunately they are not suitable for our purpose.

I noticed calling `insertOoxml` after a few seconds (approximately 5 seconds) works, but if you call it quickly, the problem I mentioned above occurs.

We use this method to insert the contents of another file, converting docx to ooxml. I tested `insertFileFromBase64` but it had the same problem as `insertOoxml`.

We also use `insertOoxml` to insert merge field, I tried using `cursor.insertField(Word.InsertLocation.before, Word.FieldType.mergeField, 'Foo Bar')` but it says it is not supported yet.

**Comment by xiruatms** [comment]
@aeturneus , thanks for providing such details. I can repro with all the information. Current investigation shows that there might be some unsupported content (on Word Online) in your testing document which we need further investigation. Those unsupported content led to such "NotAllowed" failure.

Given we can't download the testing document from your WOPI environment, can you post the testing document here (if no sensitive data there)? That'll be much helpful for us to understand.

**Comment by xiaoyuMS** [comment]
@SavioSantosDev I'm trying to fully grasp your issue. From what I understand, there are two behaviors:
1. When you execute at intervals of five seconds or so, the insertOoxml will success, which means the content appears when the waiting dialog is dismissed.
2. When you execute continuously, the content doesn't appear after the dialog is dismissed, as what you shared in the GIF.

Is my understanding accurate? If so, this is not a known issue. We might need your assistance to reproduce this behavior on our end. 

For the potential workaround, we may not have a clear signal to indicate when the next insertion can occur, you may try to execute the next insertion after the document is saved via checking the state of [Document.saved](https://learn.microsoft.com/en-us/javascript/api/word/word.document?view=word-js-preview#word-word-document-saved-member)

**Comment by SavioSantosDev** [comment]
@xiaoyuMS That's right. Thank you for the information as soon as possible I will do more tests. Let me know if you need anything to help analyze the problem.

**Comment by aeturneus** [comment]
@xiruatms we did some further analysis based on your comments and found that the use of content controls in form fields broke this functionality in WOPI. When form fields are removed and content controls are used without form fields, the insertOoxml function worked.

That said, it is quite slow (about 3-5 seconds total per insert/remove operation) and the loading modal loads twice when inserting (see screenshots) so there is a lot of room for improvement in this UX:

First modal: "inserting"
<img width="747" alt="Screenshot 2024-03-25 at 12 41 14 PM" src="https://github.com/OfficeDev/office-js/assets/877362/610ae83c-a5d3-475a-83a6-19743381c999">

Second modal: "waiting"
<img width="655" alt="Screenshot 2024-03-25 at 12 41 18 PM" src="https://github.com/OfficeDev/office-js/assets/877362/44dc4f7f-6fb5-47fa-b992-5bc2567ebb24">

Thanks for your work on this issue.

**Comment by RuizhiSunMS** [comment]
Close this bug since the original issue was resolved. The extra questions are related with some gaps.
For ones who get blocked by the same issue, plz feel free to re-open.

**Comment by aeturneus** [comment]
@RuizhiSunMS @nickamckenna @xiruatms 

I believe this issue should be re-opened as the same problem is occurring. We previously provided test account credentials and steps to reproduce in this issue (see my comments above). Those should still be valid.

Currently the insertOoxml instance fails with the console error:

`{ "code": "NotAllowed", "message": "The action isn't supported by Word in a browser. Check the OfficeExtension.Error.debugInfo for more information.", "errorLocation": "Body.insertOoxml", "statement": "var insertOoxml = body.insertOoxml(...);", "surroundingStatements": [ "var root = context.root;", "var body = root.body;", "// Instantiate {body}", "// >>>>>", "var insertOoxml = body.insertOoxml(...);", "// <<<<<", "// Instantiate {insertOoxml}" ], "fullStatements": [ "Please enable config.extendedErrorLogging to see full statements." ] }`

The same code works fine in desktop, but fails in WOPI when running with our add-in. Cross-posting from https://github.com/OfficeDev/office-js/issues/2840 as this seems to be a related issue.

**Comment by xiruatms** [comment]
@aeturneus ,  feel free to create a new GitHub issue to address your errors. And please provide details (testing environment, testing document, codes snippet) as much as possible in that issue so that we can follow up. Thanks.

**Comment by aeturneus** [comment]
@xiruatms I think @nickamckenna's code snippet in this thread can be used to reproduce the error, and we already provided these details ourselves through the survey response link posted earlier in this thread, however I will create a new issue as you request.
